Vocabulary Builder: Set Custom Intervals

Kelamir opened the pull request:
The following is a work in progress of an implementation of a setting which allows users to set custom intervals for Vocabulary Builder.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/10035)
<!-- Reviewable:end -->

weijiuqiao commented:
I received an email with your presumably deleted comment, in which you mentioned storing the intervals in the database. I believe the preferred method is storing them the same way as the other settings (and also providing a way to revert to the default scheme is needed). The database has complications in syncing and we'd better try not to alter it unless necessary :)

weijiuqiao commented:
Your comment didn't show up when I was posting the last comment :s
Anyways, we can review your code once you're ready. Another potential issue is If I want to have 20min as one interval, I'll have to calculate it, which is a little bit inconvenient. idk if maybe we can make it easier somehow?

Kelamir commented:
There is the issue of determining interval once user-chosen values run out. This can be dealt with using an interval modifier, an idea taken from https://docs.ankiweb.net/deck-options.html#interval-modifier . It would serve as a multiplier for current interval. Given last interval in array is 30d, interval modifier value of 2 would set the interval to 60d. I would like to add this, along with an option to change interval modifier, and reset it. 

This too allows users to determine the pacing of study - lower intervals, like 1.5 allow for more repetitions.

Kelamir commented:
Right, I will definitely need to add input validation, and to make it clear what current settings are.
Here's some updates:
![image](https://user-images.githubusercontent.com/61295121/213105824-2b560d07-8c65-4955-b83d-d76991202420.png)
Interval modifier button here

![image](https://user-images.githubusercontent.com/61295121/213106052-5486d35a-cd77-4fef-a549-3f389dd764af.png)
The menu pops up, explaining what it does.

There are several issues which I could use some advice on, that I will be focusing on today:
1) The text description is stored as one-liner, which is 200 characters. It's best to have it broken on multiple lines, but that creates another issue:
![image](https://user-images.githubusercontent.com/61295121/213106710-b58402a6-cb99-4baf-ab76-47b0d7151061.png)
![image](https://user-images.githubusercontent.com/61295121/213106851-08ccacf4-ceaa-4b07-b77c-b4a915704d14.png)
As you can see, it preserves spacing

2) Apply interval modifier. I did add interval modifier as a setting yesterday, but it is not doing anything yet. It's necessary to figure out logic for interval modifier to take over once array with intervals runs out of elements, and keep being applied.

3) After you look up a word, it gets added to vocabulary builder with interval 5m, instead of any custom interval set. 

4) User input validation
5) Let users know current settings
6) Better descriptions

weijiuqiao commented:
No. I was just applauding. I didn't really use unit tests and relied heavily on manual testing, which is not the best way and was quite tedious.

Kelamir commented:
@weijiuqiao would you mind taking a look at the last commit? Can't implement it and I'd been struggling with this one all day. I said previously

> It's necessary to figure out logic for interval modifier to take over once array with intervals runs out of elements, and keep being applied.

To reach that end, I've written this piece. item.last_interval would be used here to track the value of interval long past there's no elements in the array. 

```lua
    local interval
    if intervals[target_count] ~= nil then
        interval = intervals[target_count]
        item.last_interval = interval
    else
        interval = item.last_interval * settings.interval_modifier
        item.last_interval = interval
    end
```

If the array of intervals like [10m, 12h, 1d] isn't nil, it will use the interval from the array at the position of `target_count`, and assign this to `item.last_interval`. And if array out of bounds, it should set `interval` to the value of `item.last_interval` multiplied by interval modifier. However, that doesn't happen - item.last_interval is nil in the else clause, and throws error "attempt to perform arithmetic on field 'last_interval' (a nil value). I would expect it to retain the last value, such as 1d here, but it doesn't. Thanks for attention :)

Kelamir commented:
> After you look up a word, it gets added to vocabulary builder with interval 5m, instead of any custom interval set.

This is a problem that I'm facing. I don't know why that happens, what makes it happen, it's fairly confusing. I tried modifying
`stmt:bind(value.word, value.book_title or "", value.time, value.time + 5*60, value.time)`, which has 5m value hard-coded I think, but that doesn't affect it.

weijiuqiao commented:
I don't think you need to keep track of `last_interval`. You can calculate it each time you need it:
```
else
    interval = intervals[#intervals] * settings.interval_modifier * (target_count - #intervals) -- number of times the modifier needs to apply
    -- item.last_interval = interval
end
```
>After you look up a word, it gets added to vocabulary builder with interval 5m, instead of any custom interval set

I think you need to modify content of `function VocabularyBuilder:insertOrUpdate`.

Kelamir commented:
Thanks, your suggested code works great!

Kelamir commented:
Updated default steps to 5m, 30m, 14h, 24h. Afterwards interval modifier takes over. It's good to have some steps for the first day configured, because otherwise interval modifier would increase intervals like 5m, 10m, 20m, 40m. As a bonus, it's easier to explain default intervals now, and interval modifier has a better role than a strange feature that gets activated after 90 days passed.

I have removed the initial 5m interval. Admittedly, I don't know how to combine it with intervals users customize, and it might confuse them greatly. Could put it back if you prefer, then we could set defaults to 30m, 14h, 24h instead.

Kelamir commented:
![image](https://user-images.githubusercontent.com/61295121/213421593-9d3bf717-5799-4ef1-9a9f-99dfce5c09d7.png)
Now currently set intervals will be displayed.

Kelamir commented:
![image](https://user-images.githubusercontent.com/61295121/213436987-59ace6c5-cd1c-40c3-a412-c85175f44381.png)
Now both intervals and interval modifiers have their current setting displayed in the input field.

Kelamir commented:
https://user-images.githubusercontent.com/61295121/213479913-e738f1dd-e127-406b-bc2c-7363691fce8b.mp4

Here's a bit of a demonstration.

weijiuqiao commented:
>If I move function parseInputTime(str) outside MenuDialog:init, it doesn't work anymore, resulting in error "attempt to index local 'str' (a nil value)".

If you are calling the function with a dot like `MenuDialog.parseInputTime(inputTime)`, you should declare it using a dot. Or you can use `:` and declare with `:` if you need `self` within the function.

I suggest you to combine setting intervals and modifier in the same dialog, like how you can add cloud storage info, and maybe make use of an `Info` button to show the explanation that's quite lengthy?

Kelamir commented:
Some input validation added 
![image](https://user-images.githubusercontent.com/61295121/213676147-03270ab2-02f3-4ee1-b8a1-f3406fd7d304.png)

Kelamir commented:
Thanks for advice on functions and placing both input fields in the same spot. This is spot on.

Kelamir commented:
![image](https://user-images.githubusercontent.com/61295121/214383738-def7a2b6-b037-45bb-af48-8352a842c6a4.png)
![image](https://user-images.githubusercontent.com/61295121/214383820-2b4d3743-c647-4a06-be41-5b3fe53ceaab.png)

Added a simple message displaying help. Can be closed by pressing on it.

Kelamir commented:
I'm yet to add input validation for interval modifiers, after that could perhaps put it up for review.

Kelamir commented:
Hello, @weijiuqiao . Would you like to review the pull request? It took a while but it works decently by now. Here's a demonstration: 
![overview](https://user-images.githubusercontent.com/61295121/215287119-1a0e7780-a869-4a8e-bb0c-8f486f4212ee.gif)

Just realized a bit of a problem: the interval assigned to the word is a bit higher than in settings due to `math.random(math.floor(interval * 0.9), math.ceil(interval * 1.1))`, and it might cause a bit confusion for users("why is it 6m when I entered 5m?").

weijiuqiao commented on plugins/vocabbuilder.koplugin/main.lua:
```diff
@@ -242,6 +242,133 @@ function MenuDialog:init()
+
+                                local intervalsArray = {};
+                                if isIntervalValid then
+                                    for inputTime in string.gmatch(interval_input, "%d+[m|d|h]") do
```
You are duplicating the regex matching once in `isIntervalStringValid` and once here. Maybe do it just once?

weijiuqiao commented on plugins/vocabbuilder.koplugin/main.lua:
```diff
@@ -387,6 +514,50 @@ function MenuDialog:init()
+    return t
+end
+
+function MenuDialog.isIntervalStringValid(interval_string)
```
What if user inputs 0m or empty string?

Kelamir replied:
Hmm, fair concern. I did do 0m manually previously and then intervals wouldn't increase. Perhaps could ask for minimally 1m interval, also checking empty strings is needed, right.

Kelamir commented on plugins/vocabbuilder.koplugin/main.lua:
```diff
@@ -242,6 +242,133 @@ function MenuDialog:init()
+                        hint = _("10m 12h 1d"),
+                    },
+                    {
+                        description = "Interval modifier",
```
Interval modifier is an actual term, but we could simplify and rename it to interval multiplier, right.

Kelamir replied:
For starters, I will rename it as it's displayed to the user to interval multiplier.

Kelamir commented on plugins/vocabbuilder.koplugin/main.lua:
```diff
@@ -242,6 +242,133 @@ function MenuDialog:init()
+        callback = function()
+            -- review intervals
+            settings.default_review_intervals = {5, 30, 720, 1440}
+            settings.default_review_intervals_pretty = "5m, 30m, 12h, 24h"
```
Ah, true. This could instead be a local variable. Will see if that works.

Kelamir replied:
There's an issue with that. I use these values in both files, so that's done via the settings file. I'm not sure how to approach that otherwise.

weijiuqiao replied:
One naive way is to just use literals at both places, since the default is not supposed to change.
Another way is to store the default in `DB`(`VocabularyBuilder` in `db.lua`) so you can reference it from both files.

Kelamir replied:
Alright. For starters, let's use local variables. This would introduce duplication of the variable default_interval_modifier.

Kelamir commented on plugins/vocabbuilder.koplugin/db.lua:
```diff
@@ -216,26 +217,32 @@ function VocabularyBuilder:gotOrForgot(item, isGot)
+        intervals = settings.default_review_intervals
     end
 
+    local interval_modifier
```
Great, done.

NiLuJe commented on plugins/vocabbuilder.koplugin/db.lua:
```diff
@@ -3,6 +3,7 @@ local Device = require("device")
 local SQ3 = require("lua-ljsqlite3/init")
 local LuaData = require("luadata")
 local logger = require("logger")
+local settings = G_reader_settings:readSetting("vocabulary_builder", {enabled = false, with_context = true})
```
It... depends on what the setting is and how it is written.

If it's a table and the table itself is never completely rewritten, the reference should be stable, otherwise, a table rewrite with leave a dangling pointer (and a stale GC object).

Frenzie commented:
I wouldn't call that Lua specific exactly. It's a slightly terse ternary operator. In full form `condition and a or b`. Though of course it's not exactly like a ternary operator.

Kelamir commented on plugins/vocabbuilder.koplugin/main.lua:
```diff
@@ -387,6 +514,50 @@ function MenuDialog:init()
+    return totalMinutes
+end
+
+function MenuDialog.splitCommasToTable(input_string)
```
util.lua? Where would that file be? I can see only db.lua and main.lua in vocabulary builder. I haven't found my way around the rest of the codebase yet, aside from ui/widgets.

weijiuqiao replied:
frontend/util.lua

Kelamir commented:
I'm not sure why but this updated syntax with shortened if-else expressions(the last two commits) is breaking things in the code.  Pressing Default does not reset current values, for example. Works as intended when I reset back to the commit "[Fix trailing whitespace and a type error](https://github.com/koreader/koreader/pull/10035/commits/9b04190d9bd8f2535f3727a3a3be3803c2f57c05), so there must be something wrong with it.

NiLuJe commented:
What @Frenzie said (I'm pretty sure they compile to the exact same bytecode in fact).

Kelamir commented on plugins/vocabbuilder.koplugin/db.lua:
```diff
@@ -216,26 +217,32 @@ function VocabularyBuilder:gotOrForgot(item, isGot)
+        intervals = settings.review_intervals
     else
-        due_time = current_time + 24 * 3600 * 30 * 2 ^ (math.min(target_count - 8, 6))
+        intervals = settings.default_review_intervals
```
Yes, it is set there. I use the settings to access variables between files. Perhaps that's a wrong approach. It is not nil - here is output of db.lua code `logger.dbg("SETTINGS.DEFAULT_REVIEW_INTERVALS_IS ", settings.default_review_intervals)`:

```lua
01/31/23-23:36:21 DEBUG SETTINGS.DEFAULT_REVIEW_INTERVALS_IS  {
  5,
  30,
  720,
  1440
} --[[table: 0x7f993d680378]]
```

weijiuqiao commented:
> I'm not sure why but this updated syntax with shortened if-else expressions(the last two commits) is breaking things in the code. Pressing Default does not reset current values, for example. 

It might have something to do with the `settings` table not updated in `db.lua` when it's saved in `main.lua`?

Kelamir commented:
> I received an email with your presumably deleted comment, in which you mentioned storing the intervals in the database. I believe the preferred method is storing them the same way as the other settings (and also providing a way to revert to the default scheme is needed). The database has complications in syncing and we'd better try not to alter it unless necessary :)

To be clear, I have had this in mind when deciding to use settings file for everything. But it looks like there are certain complications included with that approach, thus I will try out solving interfile communication using the database.

weijiuqiao commented:
>I will try out solving interfile communication using the database.

Please don't modify the database unless necessary. The migration could be very complicated.
Do it this way:
>Another way is to store the default in DB(VocabularyBuilder in db.lua) so you can reference it from both files.

which means in the `VocabularyBuilder` table/object in `db.lua` that's been imported as `DB` in `main.lua`.

Kelamir commented:
Alright, the commits related to work on the database will be reverted. It appeared to be the only way to go forward as I had perceived a push back on the idea of using a settings file, despite it working as intended(I think, but I might be wrong). 

I understand that it might be taking a bit too long and I apologize for any inconvenience. Thanks for patience.

Kelamir commented on plugins/vocabbuilder.koplugin/db.lua:
```diff
@@ -216,26 +217,32 @@ function VocabularyBuilder:gotOrForgot(item, isGot)
+        interval = intervals[#intervals] * interval_modifier * (target_count - #intervals) -- number of times the modifier needs to apply
+    end
+
+    local interval_randomized_minutes = math.random(math.floor(interval * 0.9), math.ceil(interval * 1.1))
```
I tried this out: `local randomized_interval = item.target_count >= 2 and math.random(math.floor(interval * 0.9), math.ceil(interval * 1.1)) or 0`, got error "attempt to compare number with nil". Will look into it more tomorrow.

Kelamir commented on plugins/vocabbuilder.koplugin/db.lua:
```diff
@@ -287,8 +294,8 @@ function VocabularyBuilder:insertOrUpdate(entry)
                         next_context = ifnull(excluded.next_context, next_context);]]);
-    stmt:bind(entry.word, entry.book_title, entry.time, entry.time+300, entry.time,
-              entry.prev_context, entry.next_context, entry.time+300)
+    stmt:bind(entry.word, entry.book_title, entry.time, entry.time, entry.time,
```
I think that's the intended effect. Intervals begin applying as you open the application and start studying there. But I might be wrong.

Kelamir replied:
However, let's stick to your vision here. Will update as requested.

Kelamir commented:
Hi! Sorry for the wait. I'm going to continue working on this one shortly. The priority now is adding tests to ensure the new functionality works as intended, and fixing any bugs. I have only recently managed to get testing running(yay!).

Kelamir commented:
Closing this PR for now. I will possibly either reopen it or start a new one with the finished version at a later date. Thanks for cooperation.
